Dependent Attributes

ccfz opened the pull request.

ccfz committed:
Add lodash to dependencies

ccfz committed:
Add validation for attribute values

The only type attributes values can have in the constructor are:
functions, booleans, and strings. All other types should throw an
error.

ccfz committed:
Add support for data refs. in its own scope

The object fabricator should be able to build attributes that are
based on other values it generated in the process. For example
the e-mail attribute would build the e-mail address with the
already build first and last name

croaker commented on __tests__/index.spec.js:
```diff
@@ -76,8 +80,13 @@ describe('#constructor', () => {
+    expect(invalidNoneStringValue)
+      .toThrowError(
+        'Attribute values can only be'
+        + ' functions, strings or booleans, but received: number'
```
I don't understand why this is necessary. Can you elaborate?

croaker commented on __tests__/index.spec.js:
```diff
@@ -169,6 +178,19 @@ describe('#create', () => {
+
+    const userWithAuthor = userFabricator.create();
+
+    expect(userWithAuthor.author).toEqual('Test User 1');
```
This took me a while to understand: The 'previous' is what led me in the wrong direction. A better test description would have been e.g. 'it allows basing attribute values on other attribute values' and two attributes `name` and `author` would have been enough to make the point.

Apart from that, why did you use templates for this instead of functions (like in the sequences)? In my opinion this would lead to a more consistent API.

croaker commented on src/index.js:
```diff
@@ -8,6 +10,18 @@ export function validateAttributes(attributes) {
+
+  Object.values(attributes).forEach((attrValue) => {
+    const attrType = typeof attrValue;
+    let pass;
```
This var is unused, isn't it? You can add eslint to the project to help with these things.

croaker reviewed (commented):
Hey @ccfz, I think this is looking good, however, there are some open questions.

ccfz committed:
Revert changes to attribute value validation

ccfz committed:
Change the way attr can be based on one another

Currently their is a dependency on lodash's template base attributes
on one another. In addition the interface to use template is
different to the other special attributes.

By introducing the static function "template" the dependency on
lodash is no longer needed and a coherent interface is created.

ccfz committed:
Remove lodash dependency

ccfz commented:
@croaker thoughts?

croaker reviewed (commented):
I‘m not sure about template attributes and I would have preferred that dynamic attributes work like this:

```
userFabricator = buildTestUserFabricator({
    name: ObjectFabricator.sequence(n => `TestUser${n}`),
    email: ({ name }) => `${name}@gmail.com`
});
```

I understand this requires some refactoring. If you can think of a reasonable solution let‘s change it. Otherwise, we will live with the consequences for now and come back at a later time. We can keep the `ObjectFabricator.template` as legacy anyway.

ccfz committed:
Remove obsolete attribute type validation

ccfz commented on __tests__/index.spec.js:
```diff
@@ -169,6 +169,17 @@ describe('#create', () => {
+  it('can handle template attributes', () => {
+    userFabricator = buildTestUserFabricator({
+      name: ObjectFabricator.sequence(n => `TestUser${n}`),
+      email: ObjectFabricator.template(n => `${n.name}@gmail.com`)
```
I can think of a couple of solutions but none that are as clear and stable. The best alternative, which I built before but then got rid off because it was less clear and different to the other extensions, is using regex to match for ${...} and to replace it with ${this.filledAttributes....}. 

When filling the attributes I would then have to check if the attribute is a string and then run the regex. 

Do you think that would be better?

ccfz committed:
Refactor template spec

croaker reviewed (approved):
Plz bump the version to '0.4.0' before merging :)

ccfz committed:
Bump version to 0.4.0
